File: mage/src/main/java/mil/nga/giat/mage/map/cache/MapDataManager.java
package mil.nga.giat.mage.map.cache;

import android.app.Application;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.LifecycleRegistry;
import android.arch.lifecycle.Observer;
import android.os.AsyncTask;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.android.gms.maps.GoogleMap;

import java.io.File;
import java.net.URI;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.Executor;


@MainThread
public class MapDataManager implements LifecycleOwner {

    /**
     * Implement this interface and {@link #addUpdateListener(MapDataListener) register}
     * an instance to receive {@link #onMapDataUpdated(MapDataUpdate) notifications} when the set of mapData changes.
     */
    public interface MapDataListener {
        void onMapDataUpdated(MapDataUpdate update);
    }

    /**
     * The create update permission is an opaque interface that enforces only holders of
     * the the permission instance have the ability to create a {@link MapDataUpdate}
     * associated with a given instance of {@link MapDataManager}.  This can simply be an
     * anonymous implementation created at the call site of the {@link Config#updatePermission(CreateUpdatePermission) configuration}.
     * For example:
     * <p>
     * <pre>
     * new MapDataManager(new MapDataManager.Config()<br>
     *     .updatePermission(new MapDataManager.CreateUpdatePermission(){})
     *     // other config items
     *     );
     * </pre>
     * </p>
     * This prevents the programmer error of creating update objects outside of the
     * <code>MapDataManager</code> instance to {@link MapDataListener#onMapDataUpdated(MapDataUpdate) deliver}
     * to listeners.
     */
    public interface CreateUpdatePermission {};

    public final class MapDataUpdate {
        public final Set<MapDataResource> added;
        public final Set<MapDataResource> updated;
        public final Set<MapDataResource> removed;
        public final MapDataManager source = MapDataManager.this;

        public MapDataUpdate(CreateUpdatePermission updatePermission, Set<MapDataResource> added, Set<MapDataResource> updated, Set<MapDataResource> removed) {
            if (updatePermission != source.updatePermission) {
                throw new Error("erroneous attempt to create update from cache manager instance " + MapDataManager.this);
            }
            this.added = added;
            this.updated = updated;
            this.removed = removed;
        }
    }

    private static final String LOG_NAME = MapDataManager.class.getName();

    private static MapDataManager instance = null;

    public static class Config {
        private Application context;
        private CreateUpdatePermission updatePermission;
        private MapDataRepository[] repositories;
        private MapDataProvider[] providers;
        private Executor executor = AsyncTask.THREAD_POOL_EXECUTOR;

        public Config context(Application x) {
            context = x;
            return this;
        }

        public Config updatePermission(CreateUpdatePermission x) {
            updatePermission = x;
            return this;
        }

        public Config repositories(MapDataRepository... x) {
            repositories = x;
            return this;
        }

        public Config providers(MapDataProvider... x) {
            providers = x;
            return this;
        }

        public Config executor(Executor x) {
            executor = x;
            return this;
        }
    }

    public static synchronized void initialize(Config config) {
        if (instance == null) {
            instance = new MapDataManager(config);
            return;
        }
        throw new Error("attempt to initialize " + MapDataManager.class + " singleton more than once");
    }

    public static MapDataManager getInstance() {
        return instance;
    }

    private final CreateUpdatePermission updatePermission;
    private final Executor executor;
    private final MapDataRepository[] repositories;
    private final MapDataProvider[] providers;
    private final Collection<MapDataListener> cacheOverlayListeners = new ArrayList<>();
    private final LifecycleRegistry lifecycle = new LifecycleRegistry(this);
    private Set<MapDataResource> mapData = Collections.emptySet();
    private ImportResourcesTask importResourcesForRefreshTask;

    public MapDataManager(Config config) {
        if (config.updatePermission == null) {
            throw new IllegalArgumentException("update permission object must be non-null");
        }
        updatePermission = config.updatePermission;
        executor = config.executor;
        repositories = config.repositories;
        providers = config.providers;
        lifecycle.markState(Lifecycle.State.CREATED);
        for (MapDataRepository repo : repositories) {
            repo.observe(this, new RepositoryObserver(repo));
        }
    }

    public void addUpdateListener(MapDataListener listener) {
        cacheOverlayListeners.add(listener);
    }

    public void removeUpdateListener(MapDataListener listener) {
        cacheOverlayListeners.remove(listener);
    }

    public void tryImportResource(URI resourceUri) {
        new ImportResourcesTask().executeOnExecutor(executor, resource);
    }

    public void removeCacheOverlay(String name) {
        // TODO: rename to delete, implement MapDataProvider.deleteCache()
    }

    public Set<MapDataResource> getMapData() {
        return mapData;
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return lifecycle;
    }

    /**
     * Discover new mapData available in standard {@link MapDataRepository locations}, then remove defunct mapData.
     * Asynchronous notifications to {@link #addUpdateListener(MapDataListener) listeners}
     * will result, one notification per refresh, per listener.  Only one refresh can be active at any moment.
     */
    public void refreshAvailableCaches() {
        for (MapDataRepository repo : repositories) {
            repo.refreshAvailableMapData(executor);
        }
    }

    public MapLayerManager createMapManager(GoogleMap map) {
        return new MapLayerManager(this, Arrays.asList(providers), map);
    }

    private void onMapDataChanged(Set<MapDataResource> data, MapDataRepository source) {

    }

    private void cacheFileImportFinished(ImportResourcesTask task) {
        if (task == importResourcesForRefreshTask) {
            if (refreshTask == null) {
                throw new IllegalStateException("import task for refresh finished but refresh task is null");
            }
            refreshTask.executeOnExecutor(executor, mapData.toArray(new MapDataResource[mapData.size()]));
        }
        else {
            updateCaches(task, null);
        }
    }

    private void refreshFinished(RefreshAvailableResourcesTask task) {
        if (task != refreshTask) {
            throw new IllegalStateException(RefreshAvailableResourcesTask.class.getSimpleName() + " task completed but did not match stored task");
        }

        ImportResourcesTask localImportTask = importResourcesForRefreshTask;
        RefreshAvailableResourcesTask localRefreshTask = refreshTask;
        importResourcesForRefreshTask = null;
        refreshTask = null;

        updateCaches(localImportTask, localRefreshTask);
    }

    private void updateCaches(ImportResourcesTask importTask, RefreshAvailableResourcesTask refreshTask) {
        Set<MapDataResource> allIncoming;
        try {
            CacheImportResult importResult = importTask.get();
            allIncoming = importResult.imported;
            if (refreshTask != null) {
                allIncoming.addAll(refreshTask.get());
            }
        }
        catch (Exception e) {
            throw new IllegalStateException("unexpected error retrieving cache update results", e);
        }

        Map<MapDataResource, MapDataResource> incomingIndex = new HashMap<>();
        for (MapDataResource cache : allIncoming) {
            incomingIndex.put(cache, cache);
        }
        Set<MapDataResource> added = new HashSet<>(allIncoming);
        added.removeAll(mapData);
        Set<MapDataResource> removed = new HashSet<>();
        Set<MapDataResource> updated = new HashSet<>();
        for (MapDataResource existing : mapData) {
            MapDataResource incoming = incomingIndex.get(existing);
            if (incoming == null) {
                removed.add(existing);
            }
            else if (incoming != existing) {
                updated.add(incoming);
            }
        }

        mapData = Collections.unmodifiableSet(new HashSet<>(incomingIndex.keySet()));

        MapDataUpdate update = new MapDataUpdate(
            updatePermission,
            Collections.unmodifiableSet(added),
            Collections.unmodifiableSet(updated),
            Collections.unmodifiableSet(removed));
        for (MapDataListener listener : cacheOverlayListeners) {
            listener.onMapDataUpdated(update);
        }
    }

    private class RepositoryObserver implements Observer<Set<MapDataResource>> {

        private final MapDataRepository repo;

        private RepositoryObserver(MapDataRepository repo) {
            this.repo = repo;
        }

        @Override
        public void onChanged(@Nullable Set<MapDataResource> data) {
            onMapDataChanged(data, repo);
        }
    }

    private static class CacheImportResult {
        private final Set<MapDataResource> imported;
        // TODO: propagate failed imports to user somehow
        private final List<MapDataImportException> failed;

        private CacheImportResult(Set<MapDataResource> imported, List<MapDataImportException> failed) {
            this.imported = imported;
            this.failed = failed;
        }
    }

    private class ImportResourcesTask extends AsyncTask<MapDataResource, Void, CacheImportResult> {

        private MapDataResource importFromFirstCapableProvider(MapDataResource resource) throws MapDataImportException {
            URI uri = resource.getUri();
            for (MapDataProvider provider : providers) {
                if (uri.getScheme().equalsIgnoreCase("file")) {
                    File cacheFile = new File(uri.getPath());
                    if (!cacheFile.canRead()) {
                        throw new MapDataImportException(uri, "cache file is not readable or does not exist: " + cacheFile.getName());
                    }
                }
                if (provider.canHandleResource(uri)) {
                    return provider.importResource(uri);
                }
            }
            throw new MapDataImportException(uri, "no cache provider could handle file " + resource);
        }

        @Override
        protected CacheImportResult doInBackground(MapDataResource... resources) {
            Set<MapDataResource> caches = new HashSet<>(resources.length);
            List<MapDataImportException> fails = new ArrayList<>(resources.length);
            for (MapDataResource resource : resources) {
                MapDataResource imported;
                try {
                    imported = importFromFirstCapableProvider(resource);
                    caches.add(imported);
                }
                catch (MapDataImportException e) {
                    fails.add(e);
                }
            }
            return new CacheImportResult(caches, fails);
        }

        @Override
        protected void onPostExecute(CacheImportResult result) {
            cacheFileImportFinished(this);
        }
    }
}

File: mage/src/main/java/mil/nga/giat/mage/map/cache/MapDataManager.kt
package mil.nga.giat.mage.map.cache

import android.app.Application
import android.arch.lifecycle.*
import android.arch.lifecycle.Observer
import android.os.AsyncTask
import android.support.annotation.MainThread
import mil.nga.giat.mage.data.Resource
import mil.nga.giat.mage.data.Resource.Status.Loading
import mil.nga.giat.mage.data.Resource.Status.Success
import mil.nga.giat.mage.data.UniqueAsyncTaskManager
import java.io.File
import java.net.URI
import java.util.*
import java.util.concurrent.Executor
import kotlin.collections.HashMap
import kotlin.collections.HashSet
import kotlin.collections.LinkedHashMap


/**
 * Return a new Map excluding the pairs from this Map for which the given lambda returns true.  Return this
 * Map if the given lambda returns false for all of this Map's pairs.
 */
fun <K, V> Map<K, V>.removeIf(predicate: (K, V) -> Boolean, removeTo: MutableMap<K, V>? = null): Map<K, V> {
    val result = HashMap<K, V>()
    forEach { key, value ->
        if (!predicate.invoke(key, value)) {
            result[key] = value
        }
        else {
            removeTo?.put(key, value)
        }
    }
    if (result.size == size) {
        return this
    }
    return result
}


@MainThread
class MapDataManager(config: Config) : LifecycleOwner {

    val providers: Map<Class<out MapDataProvider>, MapDataProvider>
    // TODO: this will probably become public eventually
    private val repositories: List<MapDataRepository>
    private val executor: Executor
    private val lifecycle = LifecycleRegistry(this)
    private val mutableMapData = MutableLiveData<Resource<Map<URI, MapDataResource>>>()
    private val repositoryChangeTasks: UniqueAsyncTaskManager<String, Pair<MapDataResource, MapDataResolveException?>, ResolveRepositoryChangeResult>
    private val repositoryChangeListener = object : UniqueAsyncTaskManager.TaskListener<String, Pair<MapDataResource, MapDataResolveException?>, ResolveRepositoryChangeResult> {

        override fun taskProgress(key: String, task: UniqueAsyncTaskManager.Task<Pair<MapDataResource, MapDataResolveException?>, ResolveRepositoryChangeResult>, progress: Pair<MapDataResource, MapDataResolveException?>) {
            val change = task as ResolveRepositoryChangeTask
            change.unresolvedRemaining.remove(progress.first)
            if (progress.second == null) {
                change.result.resolved[progress.first.uri] = progress.first
            }
            else {
                change.result.failed[progress.first] = progress.second!!
            }
        }

        override fun taskCancelled(key: String, task: UniqueAsyncTaskManager.Task<Pair<MapDataResource, MapDataResolveException?>, ResolveRepositoryChangeResult>, result: ResolveRepositoryChangeResult?) {
            onResolveChangeCancelled(task as ResolveRepositoryChangeTask)
        }

        override fun taskFinished(key: String, task: UniqueAsyncTaskManager.Task<Pair<MapDataResource, MapDataResolveException?>, ResolveRepositoryChangeResult>, result: ResolveRepositoryChangeResult?) {
            onResolveFinished(task as ResolveRepositoryChangeTask, result!!)
        }
    }


    val mapData: LiveData<Resource<Map<URI, MapDataResource>>> = mutableMapData

    fun requireMapData(): Resource<Map<URI, MapDataResource>> { return mutableMapData.value!! }
    /**
     * Return a non-null map of resources, keyed by their [URIs][MapDataResource.uri].
     */
    var resources: Map<URI, MapDataResource> = emptyMap()
        private set
    /**
     * Return a non-null, mutable map of all layers from every [resource][resources], keyed by [layer URI][MapLayerDescriptor.layerUri].
     * Changes to the returned mutable map have no effect on this [MapDataManager]'s layers and resources.
     */
    val layers: MutableMap<URI, MapLayerDescriptor> get() = resources.values.flatMap { it.layers.values }.associateBy { it.layerUri }.toMutableMap()

    init {
        providers = Collections.unmodifiableMap(config.providers!!.associateByTo(LinkedHashMap()) { it.javaClass })
        repositories = config.repositories!!.asList()
        executor = config.executor!!
        lifecycle.markState(Lifecycle.State.RESUMED)
        mutableMapData.value = Resource.success(emptyMap())
        repositoryChangeTasks = UniqueAsyncTaskManager(repositoryChangeListener, executor)
        for (repo in repositories) {
            repo.observe(this, RepositoryObserver(repo))
        }
    }

    fun resourceForLayer(layer: MapLayerDescriptor): MapDataResource? {
        return resources[layer.resourceUri]
    }

    /**
     * Attempt to import the data the given resource URI references.  If a potential path to importing the data
     * exists, begin the asynchronous import process.
     *
     * @return true if an async import operation will begin, false if the resource cannot be imported
     */
    fun tryImportResource(resourceUri: URI): Boolean {
        for (repo in repositories) {
            if (repo.ownsResource(resourceUri)) {
                repo.refreshAvailableMapData(resourcesForRepo(repo), executor)
                return true
            }
        }
        return false
    }

    override fun getLifecycle(): Lifecycle {
        return lifecycle
    }

    /**
     * Discover new resources available in known [locations][MapDataRepository], then remove defunct resources.
     * One or more asynchronous notifications to [observers][mapData] will result as data from the various
     * [repositories][MapDataRepository] loads and [resolves][MapDataProvider.resolveResource].
     */
    fun refreshMapData() {
        for (repo in repositories) {
            if (repo.value?.status != Loading && !repositoryChangeTasks.isRunningTaskForKey(repo.id)) {
                repo.refreshAvailableMapData(resourcesForRepo(repo), executor)
            }
        }
    }

    private fun resourcesForRepo(repo: MapDataRepository): Map<URI, MapDataResource> {
        return resources.filter { it.value.repositoryId == repo.id }
    }

    private fun onMapDataChanged(resource: Resource<Set<MapDataResource>>?, source: MapDataRepository) {
        val preChangeSize = requireMapData().content!!.size
        val preChangeStatus = requireMapData().status
        if (resource?.status == Loading) {
            resources = resources.removeIf({ _, res ->
                res.repositoryId == source.id
            })
            if (resources.size < preChangeSize || preChangeStatus != Loading) {
                mutableMapData.value = Resource.loading(resources)
            }
            return
        }

        val change = resource?.content ?: emptySet()
        val unchanged = HashSet<URI>()
        val resolved = HashMap<URI, MapDataResource>(change.size)
        val unresolved = HashMap<URI, MapDataResource>(change.size)
        for (changeRes in change) {
            when {
                changeRes.resolved == null -> unresolved[changeRes.uri] = changeRes
                resources[changeRes.uri] !== changeRes -> resolved[changeRes.uri] = changeRes
                else -> unchanged.add(changeRes.uri)
            }
        }
        val remove = HashMap<URI, MapDataResource>()
        for ((uri, existing) in resources) {
            if (existing.repositoryId == source.id && !resolved.containsKey(uri) && !unchanged.contains(uri)) {
                remove[uri] = existing
            }
        }

        var nextStatus = Success
        when {
            unresolved.isNotEmpty() -> {
                nextStatus = Loading
                repositoryChangeTasks.execute(source.id, ResolveRepositoryChangeTask(source, unresolved, resourcesForRepo(source), ArrayList(providers.values)))
            }
            repositoryChangeTasks.isRunningTaskForKey(source.id) -> {
                if (repositoryChangeTasks.pendingTaskForKey(source.id) != null) {
                    nextStatus = Loading
                }
                repositoryChangeTasks.cancelTasksForKey(source.id)
            }
            repositories.any { it.value?.status == Loading } -> nextStatus = Loading
        }

        if (nextStatus == preChangeStatus && resolved.isEmpty() && remove.isEmpty()) {
            return
        }

        resources -= remove.keys
        resources += resolved

        mutableMapData.value = Resource(resources, nextStatus)
    }

    private fun onResolveFinished(change: ResolveRepositoryChangeTask, result: ResolveRepositoryChangeResult) {
        change.repository.onExternallyResolved(HashSet(result.resolved.values))
    }

    private fun onResolveChangeCancelled(cancelled: ResolveRepositoryChangeTask) {
        val pending = repositoryChangeTasks.pendingTaskForKey(cancelled.repository.id) as ResolveRepositoryChangeTask?
//        if (pending == null) {
//            onMapDataChanged(cancelled.repository.value, cancelled.repository)
//        }
//        else {
            pending?.updateProgressFromCancelledChange(cancelled)
//        }
    }

    class Config {

        var context: Application? = null
            private set
        var repositories: Array<out MapDataRepository>? = emptyArray()
            private set
        var providers: Array<out MapDataProvider>? = emptyArray()
            private set
        var executor: Executor? = AsyncTask.THREAD_POOL_EXECUTOR
            private set

        fun context(x: Application): Config {
            context = x
            return this
        }

        fun repositories(vararg x: MapDataRepository): Config {
            repositories = x
            return this
        }

        fun providers(vararg x: MapDataProvider): Config {
            providers = x
            return this
        }

        fun executor(x: Executor): Config {
            executor = x
            return this
        }
    }

    private inner class RepositoryObserver internal constructor(private val repo: MapDataRepository) : Observer<Resource<Set<MapDataResource>>> {

        override fun onChanged(data: Resource<Set<MapDataResource>>?) {
            onMapDataChanged(data, repo)
        }
    }

    private class ResolveRepositoryChangeTask(
            val repository: MapDataRepository,
            unresolved: Map<URI, MapDataResource>,
            private val existingResolved: Map<URI, MapDataResource>,
            private val providers: List<MapDataProvider>)
        : UniqueAsyncTaskManager.Task<Pair<MapDataResource, MapDataResolveException?>, ResolveRepositoryChangeResult> {

        val unresolvedRemaining: SortedSet<MapDataResource> = TreeSet { a, b -> a.uri.compareTo(b.uri) }
        val result = ResolveRepositoryChangeResult()
        val stringValue: String
        var backgroundUnresolved: Array<MapDataResource>

        init {
            for (resource in unresolved.values) {
                unresolvedRemaining.add(resource)
            }
            backgroundUnresolved = unresolvedRemaining.toTypedArray()
            stringValue = "${javaClass.simpleName} repo ${repository.id} resolving\n  ${unresolvedRemaining.map { it.uri }}"
        }


        @Throws(MapDataResolveException::class)
        private fun resolveWithFirstCapableProvider(resource: MapDataResource): MapDataResource {
            val uri = resource.uri
            if (uri.scheme.equals("file", ignoreCase = true)) {
                val resourceFile = File(uri)
                if (!resourceFile.canRead()) {
                    throw MapDataResolveException(uri, "resource file is not readable or does not exist: ${resourceFile.name}")
                }
            }
            val existing = existingResolved[uri]
            if (existing != null && existing.contentTimestamp >= resource.contentTimestamp) {
                return resource.resolve(existing.resolved!!)
            }
            for (provider in providers) {
                if (provider.canHandleResource(resource)) {
                    return provider.resolveResource(resource)
                }
            }
            throw MapDataResolveException(uri, "no cache provider could handle resource $resource")
        }

        override fun run(support: UniqueAsyncTaskManager.TaskSupport<Pair<MapDataResource, MapDataResolveException?>>): ResolveRepositoryChangeResult? {
            backgroundUnresolved.forEach { unresolved ->
                if (support.isCancelled()) {
                    return@forEach
                }
                try {
                    val resolved = resolveWithFirstCapableProvider(unresolved)
                    // if cancelled while this resolve happens, AsyncTask will not publish the progress,
                    // so capture it anyway to avoid potentially doing unnecessary work again in the next
                    // resolve for the associated repository
                    support.reportProgressToMainThread(Pair(resolved, null))
                }
                catch (e: MapDataResolveException) {
                    support.reportProgressToMainThread(Pair(unresolved, e))
                }
            }
            return result
        }

        @MainThread
        fun updateProgressFromCancelledChange(cancelledChange: MapDataManager.ResolveRepositoryChangeTask) {
            val cursor = unresolvedRemaining.iterator()
            while (cursor.hasNext()) {
                val unresolvedResource = cursor.next()
                val resolvedResource = cancelledChange.result.resolved[unresolvedResource.uri]
                if (resolvedResource != null && resolvedResource.contentTimestamp >= unresolvedResource.contentTimestamp) {
                    result.resolved[unresolvedResource.uri] = unresolvedResource.resolve(resolvedResource.resolved!!)
                    cursor.remove()
                }
            }
            backgroundUnresolved = unresolvedRemaining.toTypedArray()
        }

        override fun toString(): String {
            return stringValue
        }
    }


    private class ResolveRepositoryChangeResult {

        val resolved = HashMap<URI, MapDataResource>()
        // TODO: propagate failed imports to user somehow
        val failed = HashMap<MapDataResource, MapDataResolveException>()
    }

    companion object {

        @JvmStatic
        var instance: MapDataManager? = null
            private set

        @JvmStatic
        @Synchronized
        fun initialize(config: Config) {
            if (MapDataManager.instance != null) {
                throw Error("attempt to initialize " + MapDataManager::class.java + " singleton more than once")
            }
            instance = MapDataManager(config)
        }
    }
}
